library/alloc/src/str.rs

```diff
@@ -615,50 +603,80 @@ pub unsafe fn from_boxed_utf8_unchecked(v: Box<[u8]>) -> Box<str> {
+/// Converts leading ascii bytes in `s` by calling the `convert` function.
+///
 /// For better average performance, this happens in chunks of `2*size_of::<usize>()`.
-/// Returns a vec with the converted bytes.
+///
+/// Returns a tuple of the converted prefix and the remainder starting from
+/// the first non-ascii character.
 #[inline]
 #[cfg(not(test))]
 #[cfg(not(no_global_oom_handling))]
-fn convert_while_ascii(b: &[u8], convert: fn(&u8) -> u8) -> Vec<u8> {
-    let mut out = Vec::with_capacity(b.len());
-
+fn convert_while_ascii(s: &str, convert: fn(&u8) -> u8) -> (String, &str) {
+    // process the input in chunks to enable auto-vectorization
     const USIZE_SIZE: usize = mem::size_of::<usize>();
     const MAGIC_UNROLL: usize = 2;
     const N: usize = USIZE_SIZE * MAGIC_UNROLL;
-    const NONASCII_MASK: usize = usize::from_ne_bytes([0x80; USIZE_SIZE]);
 
-    let mut i = 0;
-    unsafe {
-        while i + N <= b.len() {
-            // Safety: we have checks the sizes `b` and `out` to know that our
-            let in_chunk = b.get_unchecked(i..i + N);
-            let out_chunk = out.spare_capacity_mut().get_unchecked_mut(i..i + N);
-
-            let mut bits = 0;
-            for j in 0..MAGIC_UNROLL {
-                // read the bytes 1 usize at a time (unaligned since we haven't checked the alignment)
-                // safety: in_chunk is valid bytes in the range
-                bits |= in_chunk.as_ptr().cast::<usize>().add(j).read_unaligned();
-            }
-            // if our chunks aren't ascii, then return only the prior bytes as init
-            if bits & NONASCII_MASK != 0 {
-                break;
-            }
+    let mut slice = s.as_bytes();
+    let mut out = Vec::with_capacity(slice.len());
+    let mut out_slice = &mut out.spare_capacity_mut()[..slice.len()];
 
-            // perform the case conversions on N bytes (gets heavily autovec'd)
-            for j in 0..N {
-                // safety: in_chunk and out_chunk is valid bytes in the range
-                let out = out_chunk.get_unchecked_mut(j);
-                out.write(convert(in_chunk.get_unchecked(j)));
-            }
+    let mut ascii_prefix_len = 0_usize;
+    let mut is_ascii = [false; N];
+
+    while slice.len() >= N {
+        // Safety: checked in loop condition
+        let chunk = unsafe { slice.get_unchecked(..N) };
+        // Safety: out_slice has same length as input slice and gets sliced with the same offsets
+        let out_chunk = unsafe { out_slice.get_unchecked_mut(..N) };
+
+        for j in 0..N {
+            is_ascii[j] = chunk[j] <= 127;
+        }
 
-            // mark these bytes as initialised
-            i += N;
+        // auto-vectorization for this check is a bit fragile,
+        // sum and comparing against the chunk size gives the best result,
+        // specifically a pmovmsk instruction on x86.
+        if is_ascii.iter().map(|x| *x as u8).sum::<u8>() as usize != N {
+            break;
         }
```

Comment:
Ah sorry, on more thing, I thought I had already mentioned this. If it is fragile it'd be good to have a codegen or assembly test for it.

Reply:
Good idea, trying this right now, but ran into an issue. All codegen tests I've seen so far are for `inline` functions, so llvm ir just contains a `tailcall`. Haven't tried assembly yet, but I guess it would need to match on the manged name of `str::to_lowercase`. Do you know any existing tests for non-inline functions?

Reply:
That complicates things. At least on godbolt I can get it to inline `to_lowercase` in the assembly when building a binary with `-Clto`. Looks like another PR (https://github.com/rust-lang/rust/pull/122770#issuecomment-2016574563) recently ran into the same limitation and they couldn't get codegen tests to work either.

So it probably has to be an assembly test for now.

Reply:
This assembly test appears to inline everything afaict

```rust
//@ assembly-output: emit-asm
//@ compile-flags: --crate-type bin -Cprefer-dynamic=no -O -Cdebuginfo=0 -Cpanic=abort -Clto

use std::hint::black_box;

#[inline]
fn lower(s: &str) -> String {
    // CHECK: fail-the-test
    s.to_lowercase()
}

fn main() {
    black_box(lower(black_box("foo")));
}
```

Reply:
Another option is to write a simplified version that just checks the `if iter().map(|x| *x as u8).sum::<u8>() as usize != N {break}` pattern in a vectorized add loop.

Reply:
I could not get it to inline, also trying `-Cllvm-args=--inline-threshold=large_number` did not have any effect, so went with the second approach. The code could maybe be minimized a bit more, but including the lowercasing makes it more similar to the actual library code.

Reply:
@the8472 do you have any further feedback regarding the test?

I also opened an [llvm issue about recognizing more movmsk like patterns](https://github.com/llvm/llvm-project/issues/96395) .

Reply:
Test now uses `convert_while_ascii` directly via the unstable `str_internals` feature.